File: backend/tauri/src/ai_advisor.rs
use serde::{Deserialize, Serialize};
use std::env;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PackageAnalysis {
    #[serde(rename = "packageName")]
    pub package_name: String,
    pub summary: String,
    pub purpose: String,
    pub dependencies: Vec<String>,
    #[serde(rename = "safeToRemove")]
    pub safe_to_remove: bool,
    #[serde(rename = "riskCategory")]
    pub risk_category: String,
    pub consequences: Vec<String>,
    #[serde(rename = "userReports")]
    pub user_reports: Vec<String>,
    #[serde(rename = "technicalDetails")]
    pub technical_details: String,
    #[serde(rename = "bestCase")]
    pub best_case: String,
    #[serde(rename = "worstCase")]
    pub worst_case: String,
}

#[derive(Debug, Serialize)]
struct PerplexityRequest {
    model: String,
    messages: Vec<PerplexityMessage>,
    max_tokens: u32,
    temperature: f32,
    top_p: f32,
    return_citations: bool,
    return_images: bool,
    search_recency_filter: String,
}

#[derive(Debug, Serialize)]
struct PerplexityMessage {
    role: String,
    content: String,
}

#[derive(Debug, Deserialize)]
struct PerplexityResponse {
    choices: Vec<PerplexityChoice>,
}

#[derive(Debug, Deserialize)]
struct PerplexityChoice {
    message: PerplexityResponseMessage,
}

#[derive(Debug, Deserialize)]
struct PerplexityResponseMessage {
    content: String,
}

/// Analyzes an Android package using Perplexity AI
pub async fn analyze_package(package_name: &str) -> Result<PackageAnalysis, String> {
    // Get API key from environment variable
    let api_key = env::var("PERPLEXITY_API_KEY")
        .map_err(|_| "PERPLEXITY_API_KEY environment variable not set. Please add it to your .env file.".to_string())?;

    // Build the prompt
    let prompt = format!(
        r#"Analyze the Android package "{}" and provide a comprehensive safety assessment.

OUTPUT MUST BE VALID JSON with this exact structure:
{{
  "packageName": "{}",
  "summary": "One-sentence plain English explanation",
  "purpose": "Technical function and role in Android system",
  "dependencies": ["list", "of", "dependent", "packages"],
  "safeToRemove": true_or_false,
  "riskCategory": "Safe|Caution|Expert|Dangerous",
  "consequences": ["what", "breaks", "if", "removed"],
  "userReports": ["known issues from Reddit/XDA/forums"],
  "technicalDetails": "Deep technical explanation",
  "bestCase": "Best outcome if uninstalled",
  "worstCase": "Worst outcome if uninstalled"
}}

Assessment criteria:
1. **Safe**: Third-party bloatware, telemetry, ads - no system impact
2. **Caution**: Vendor features (themes, gestures) - removable but affects UX
3. **Expert**: System services with alternatives available - requires knowledge
4. **Dangerous**: Core Android components - removal causes bootloop/crash

CRITICAL RULES:
- Output ONLY valid JSON, no markdown, no code blocks, no explanations
- All string values must be actual content, not placeholders
- Dependencies must be real package names or empty array
- User reports must be specific findings or "No major issues reported"
- Risk category must be one of: Safe, Caution, Expert, Dangerous

Search Reddit, XDA, and Android forums for real user experiences with this package."#,
        package_name, package_name
    );

    // Build request
    let request_body = PerplexityRequest {
        model: "sonar".to_string(), // Valid Perplexity model
        messages: vec![PerplexityMessage {
            role: "user".to_string(),
            content: prompt,
        }],
        max_tokens: 2000,
        temperature: 0.2,
        top_p: 0.9,
        return_citations: true,
        return_images: false,
        search_recency_filter: "month".to_string(),
    };

    // Make HTTP request
    let client = reqwest::Client::new();
    let response = client
        .post("https://api.perplexity.ai/chat/completions")
        .header("Content-Type", "application/json")
        .header("Authorization", format!("Bearer {}", api_key))
        .json(&request_body)
        .send()
        .await
        .map_err(|e| format!("Failed to connect to Perplexity API: {}", e))?;

    // Check status code
    let status = response.status();
    if !status.is_success() {
        let error_text = response.text().await.unwrap_or_else(|_| "Unknown error".to_string());
        return Err(format!("Perplexity API error ({}): {}", status, error_text));
    }

    // Parse response
    let perplexity_response: PerplexityResponse = response
        .json()
        .await
        .map_err(|e| format!("Failed to parse Perplexity response: {}", e))?;

    // Extract content
    let content = perplexity_response
        .choices
        .first()
        .ok_or_else(|| "No response from Perplexity API".to_string())?
        .message
        .content
        .clone();

    // Clean JSON from potential markdown code blocks
    let json_content = if content.contains("```json") {
        content
            .split("```json")
            .nth(1)
            .and_then(|s| s.split("```").next())
            .unwrap_or(&content)
            .trim()
    } else if content.contains("```") {
        content
            .split("```")
            .nth(1)
            .and_then(|s| s.split("```").next())
            .unwrap_or(&content)
            .trim()
    } else {
        content.trim()
    };

    // Parse the AI response as JSON
    let analysis: PackageAnalysis = serde_json::from_str(json_content)
        .map_err(|e| format!("Failed to parse AI response as JSON: {}. Response was: {}", e, json_content))?;

    // Validate risk category
    match analysis.risk_category.as_str() {
        "Safe" | "Caution" | "Expert" | "Dangerous" => Ok(analysis),
        _ => {
            let mut fixed_analysis = analysis;
            fixed_analysis.risk_category = "Caution".to_string();
            Ok(fixed_analysis)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    #[ignore] // Run manually with: cargo test -- --ignored
    async fn test_analyze_package() {
        // Set API key for testing
        env::set_var("PERPLEXITY_API_KEY", "your-test-key-here");
        
        let result = analyze_package("com.miui.analytics").await;
        match result {
            Ok(analysis) => {
                println!("Analysis: {:#?}", analysis);
                assert!(!analysis.summary.is_empty());
                assert!(["Safe", "Caution", "Expert", "Dangerous"]
                    .contains(&analysis.risk_category.as_str()));
            }
            Err(e) => {
                println!("Error: {}", e);
            }
        }
    }
}

File: backend-python/ai_advisor.py
"""
AI Advisor Module
Handles AI-powered package analysis using Perplexity or OpenAI
"""
import os
import json
import requests
from typing import Dict
from dotenv import load_dotenv

# Load environment variables
load_dotenv()


class AIAdvisor:
    """AI-powered package analysis"""
    
    def __init__(self, provider: str = "perplexity"):
        """
        Initialize AI advisor
        
        Args:
            provider: "perplexity" or "openai"
        """
        self.provider = provider
        
        if provider == "perplexity":
            self.api_key = os.getenv("PERPLEXITY_API_KEY")
            self.api_url = "https://api.perplexity.ai/chat/completions"
            self.model = "sonar"  # Updated model name for Perplexity API
        else:  # openai
            self.api_key = os.getenv("OPENAI_API_KEY")
            self.api_url = "https://api.openai.com/v1/chat/completions"
            self.model = "gpt-4-turbo-preview"
        
        if not self.api_key:
            raise ValueError(f"{provider.upper()}_API_KEY not found in environment variables")
    
    def analyze_package(self, package_name: str) -> Dict:
        """Analyze an Android package and return safety information"""
        
        prompt = f"""You are an Android package analysis expert. Analyze package: {package_name}

Return ONLY valid JSON (no markdown, no explanation):
{{
  "packageName": "{package_name}",
  "summary": "Brief description in plain English",
  "purpose": "What this package does",
  "dependencies": ["list of packages that might depend on this"],
  "safeToRemove": true/false,
  "riskCategory": "Safe/Caution/Expert/Dangerous",
  "consequences": ["what happens if removed"],
  "userReports": ["common user experiences"],
  "technicalDetails": "Technical information",
  "bestCase": "Best case scenario if removed",
  "worstCase": "Worst case scenario if removed"
}}

Risk categories:
- Safe: Third-party apps, easily reinstallable
- Caution: OEM apps, may affect minor features
- Expert: May break functionality
- Dangerous: Critical system components"""

        try:
            headers = {
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json"
            }
            
            messages = []
            
            # Perplexity doesn't support system role, combine into user message
            if self.provider == "perplexity":
                messages = [
                    {
                        "role": "user",
                        "content": prompt
                    }
                ]
            else:
                # OpenAI supports system role
                messages = [
                    {
                        "role": "system",
                        "content": "You are an Android debloating expert. Always respond with valid JSON only."
                    },
                    {
                        "role": "user",
                        "content": prompt
                    }
                ]
            
            payload = {
                "model": self.model,
                "messages": messages,
                "temperature": 0.2,
                "max_tokens": 1500
            }
            
            # Add provider-specific parameters
            if self.provider == "perplexity":
                payload.update({
                    "return_citations": False,
                    "return_images": False,
                    "search_recency_filter": "month"
                })
            
            response = requests.post(
                self.api_url,
                headers=headers,
                json=payload,
                timeout=30
            )
            
            # Debug: Print error details if request fails
            if response.status_code != 200:
                print(f"Error Response Status: {response.status_code}")
                print(f"Error Response Body: {response.text}")
            
            response.raise_for_status()
            data = response.json()
            
            # Extract content
            content = data["choices"][0]["message"]["content"]
            
            # Clean markdown if present
            content = content.strip()
            if content.startswith("```json"):
                content = content[7:]
            if content.startswith("```"):
                content = content[3:]
            if content.endswith("```"):
                content = content[:-3]
            content = content.strip()
            
            # Parse JSON
            analysis = json.loads(content)
            return analysis
            
        except requests.exceptions.RequestException as e:
            raise Exception(f"API request failed: {str(e)}")
        except json.JSONDecodeError as e:
            raise Exception(f"Failed to parse AI response: {str(e)}")
        except Exception as e:
            raise Exception(f"Analysis failed: {str(e)}")
    
    def chat(self, message: str, history: list = None) -> str:
        """Chat with AI about debloating"""
        
        if history is None:
            history = []
        
        try:
            headers = {
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json"
            }
            
            messages = []
            
            # Perplexity doesn't support system role
            if self.provider == "perplexity":
                # Just add history and current message
                messages.extend(history)
                messages.append({
                    "role": "user",
                    "content": message
                })
            else:
                # OpenAI supports system role
                messages = [
                    {
                        "role": "system",
                        "content": """You are a helpful Android debloating expert assistant. 
Help users understand which apps are safe to remove and answer their questions about Android packages.
Be concise and friendly."""
                    }
                ]
                # Add history
                messages.extend(history)
                # Add current message
                messages.append({
                    "role": "user",
                    "content": message
                })
            
            payload = {
                "model": self.model,
                "messages": messages,
                "temperature": 0.7,
                "max_tokens": 800
            }
            
            if self.provider == "perplexity":
                payload.update({
                    "return_citations": False,
                    "return_images": False
                })
            
            response = requests.post(
                self.api_url,
                headers=headers,
                json=payload,
                timeout=30
            )
            
            response.raise_for_status()
            data = response.json()
            
            return data["choices"][0]["message"]["content"]
            
        except Exception as e:
            raise Exception(f"Chat failed: {str(e)}")


# Example usage
if __name__ == "__main__":
    # Test the advisor
    advisor = AIAdvisor(provider="perplexity")
    
    result = advisor.analyze_package("com.facebook.katana")
    print(json.dumps(result, indent=2))
